doc/config.rst
.. highlight:: none

.. _config:

Configuration
=============

This section describes how to configure Mutalyzer and includes a list of all
available configuration settings.

Mutalyzer looks for its configuration in the file specified by the
``MUTALYZER_SETTINGS`` environment variable. Make sure to always have this
environment variable set when invoking any component of Mutalyzer. One way of
doing this is by exporting it::

    $ export MUTALYZER_SETTINGS=~/mutalyzer/settings.py

If you like, you can add this command to your ``~/.bashrc`` to have it
executed every time you open a shell.

Another way is by prefixing your invocations with
``MUTALYZER_SETTINGS=...``. For example::

    $ MUTALYZER_SETTINGS=~/mutalyzer/settings.py mutalyzer-website


Example configuration
---------------------

If you followed the steps in :ref:`install`, this is a standard configuration
file that will work for you:

.. code-block:: python

    REDIS_URI = 'redis://localhost'
    DATABASE_URI = 'postgresql://mutalyzer:*****@localhost/mutalyzer'

This is not yet a minimal configuration. In fact, you can run Mutalyzer
without a configuration file since the default configuration works out of the
box. The default configuration uses a mock Redis instance, an in-memory
database backend and a temporary directory for cache and log files, so it is
not recommended for anything more than playing around.

The next section describes all available configuration settings.


Configuration settings
----------------------

Note that the configuration file is interpreted as a Python module, so you can
use arbitrary Python expressions as configuration values, or even import other
modules in it.

Unsetting a configuration setting is done by using the value `None`. If no
default value is mentioned for any configuration setting below it means it is
not set by default.

.. _config-email:

EMAIL
  The email address used in contact information on the website and sent with
  NCBI Entrez calls.

  `Default value:` ``info@mutalyzer.nl``

BATCH_NOTIFICATION_EMAIL
  The email address used as sender in batch job notifications. If set to
  `None`, the value of :ref:`EMAIL <config-email>` will be used.

  `Default value:` `None`

.. _config-debug:

DEBUG
  If set to `True`, Mutalyzer runs in debug mode and will show more
  information with errors.

  `Default value:` `False`

.. _config-cache-dir:

CACHE_DIR
  The cache directory which is used to store uploaded and downloaded files
  such as reference files from the NCBI and batch job results.

  `Default value:` ``/tmp``


User input settings
^^^^^^^^^^^^^^^^^^^

MAX_FILE_SIZE
  Maximum size for uploaded and downloaded files (in bytes).

  `Default value:` `10 * 1048576` (10 MB)

EXTRACTOR_MAX_INPUT_LENGTH
  Maximum sequence length for description extractor (in bases).

  `Default value:` `50 * 1000` (50 Kbp)

BATCH_JOBS_ERROR_THRESHOLD
  Allow for this fraction of errors in batch jobs.

  `Default value:` `0.05`


Database settings
^^^^^^^^^^^^^^^^^

DATABASE_URI
  SQLAlchemy database connection URI specifying the database used to store
  users, samples, variants, etcetera.

  ================   =============================================================
  Database system    Example URI
  ================   =============================================================
  PostgreSQL         ``postgresql://mutalyzer:*****@localhost/mutalyzer``
  MySQL              ``mysql://mutalyzer:*****@localhost/mutalyzer?charset=utf8``
  SQLite             ``sqlite:////tmp/mutalyzer.db``
  ================   =============================================================

  See the SQLAlchemy documentation on
  `Engine Configuration
  <http://docs.sqlalchemy.org/en/latest/core/engines.html>`_ for more
  information.

  `Default value:` ``sqlite://`` (in-memory SQLite database)

REDIS_URI
  Redis connection URI (can be any `redis-py
  <https://github.com/andymccurdy/redis-py>`_ connection URI). Set to `None`
  to silently use a mock Redis. Redis is only used for non-essential
  features such as caching of external resources.

  `Default value:` `None`


Settings for output and logging
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

All Mutalyzer messages come with a level which can be one of:

======  ========  ======================================================
Level   Alias     Meaning
======  ========  ======================================================
-1      Log       Specifically log a message.
0       Debug     Debug information.
1       Info      Info.
2       Warning   Regular warnings.
3       Error     Serious errors that can be compensated for.
4       Fatal     Errors that are not recoverable.
5       Off       Can be used as a log/output level to turn off output.
======  ========  ======================================================

LOG_FILE
  Name and location of the log file.

  `Default value:` ``/tmp/mutalyzer.log``

LOG_LEVEL
  Level of logged messages.

  `Default value:` `3`

OUTPUT_LEVEL
  Level of output messages.

  `Default value:` `1`

LOG_TIME_FORMAT
  Format of time prefix for log messages. Can be anything that is accepted as
  the format argument of `time.strftime
  <http://docs.python.org/2/library/time.html#time.strftime>`_.

  `Default value:` ``%Y-%m-%d %H:%M:%S``


Website settings
^^^^^^^^^^^^^^^^

.. _config-reverse-proxied:

REVERSE_PROXIED
  If set to `True`, the WSGI application runs behind a reverse proxy (e.g.,
  nginx using ``proxy_pass``). This needs to be set if the application is
  mapped to a URL other than / or a different HTTP scheme is used by the
  reverse proxy.

  `Default value:` `False`

.. _config-website-root-url:

WEBSITE_ROOT_URL
  URL to the website root (without trailing slash). Used for generating
  download links in the batch scheduler.

  `Default value:` `None`

.. _config-soap-wsdl-url:

SOAP_WSDL_URL
  URL to the SOAP webservice WSDL document. Used to build the WSDL document
  and for linking to it from the documentation page on the website.

  `Default value:` `None`

.. _config-json-root-url:

JSON_ROOT_URL
  URL to the HTTP/RPC+JSON webservice root (without trailing slash). Used for
  linking to it from the documentation page on the website.

  `Default value:` `None`


Piwik settings
^^^^^^^^^^^^^^

`Piwik <http://piwik.org/>`_ is an Open Source analytics platform. Mutalyzer
has built-in support for visitor tracking with Piwik.

PIWIK
  If set to `True`, Piwik is enabled and some Javascript tracking code is
  included in every Mutalyzer website page.

  `Default value:` `False`

PIWIK_BASE_URL
  Base URL for the Piwik server.

  `Default value:` ``https://piwik.example.com``

PIWIK_SITE_ID
  Piwik site ID for Mutalyzer.

  `Default value:` `1`


Miscellaneous settings
^^^^^^^^^^^^^^^^^^^^^^

LRG_PREFIX_URL
  Prefix URL from where LRG files are fetched.

  `Default value:` ``ftp://ftp.ebi.ac.uk/pub/databases/lrgex/SCHEMA_1_7_ARCHIVE/``

DEFAULT_ASSEMBLY
  Default genome assembly (by name or alias).

  `Default value:` ``hg19``

NEGATIVE_LINK_CACHE_EXPIRATION
  Cache expiration time for negative transcript<->protein links from the NCBI
  (in seconds).

  `Default value:` `60 * 60 * 24 * 30` (30 days)

USE_RELOADER
  Enable the `Werkzeug reloader
  <http://werkzeug.pocoo.org/docs/0.10/serving/#reloader>`_ for the website.

  This is disabled by default due to `a bug with using the reloader
  <https://github.com/mitsuhiko/werkzeug/issues/461#issuecomment-139369694>`_
  in combination with ``python -m mutalyzer.entrypoints.website``.

  `Default value:` `False`


CHANGES.rst
Changelog
=========

This is a record of changes made between each Mutalyzer release.

Version 2.0.33
---------------

Release date to be decided.


Version 2.0.32
---------------

Released on December 9th 2019.

- Update email address and wiki link (`#498
  <https://github.com/mutalyzer/mutalyzer/pull/498>`_).
- Fix erroneous message for mtDNA m. coordinate system (`#497
  <https://github.com/mutalyzer/mutalyzer/pull/497>`_).
- Update requirements (`#496
  <https://github.com/mutalyzer/mutalyzer/pull/496>`_).
- Fix invalid LRG record issue (`#495
  <https://github.com/mutalyzer/mutalyzer/pull/495>`_).


Version 2.0.31
---------------

Released on August 21st 2019.

- Update requirements (`#486
  <https://github.com/mutalyzer/mutalyzer/pull/486>`_).
- Add support for NCBI E-utilities API key (`#485
  <https://github.com/mutalyzer/mutalyzer/pull/485>`_).
- Update HGVS descriptions retrieval for SNP converter (`#482
  <https://github.com/mutalyzer/mutalyzer/issues/482>`_)
- Make file selection mandatory in the batch job web page  (`#481
  <https://github.com/mutalyzer/mutalyzer/issues/481>`_).
- Update Travis configuration (`#483
  <https://github.com/mutalyzer/mutalyzer/pull/483>`_)


Version 2.0.30
---------------

Released on May 24th 2019.

- Update requirements (`#475
  <https://github.com/mutalyzer/mutalyzer/pull/475>`_).
- Update documentation links (`#474
  <https://github.com/mutalyzer/mutalyzer/pull/474>`_).
- Fix the genbank parser to perform loci objects creation in two steps (`#468
  <https://github.com/mutalyzer/mutalyzer/issues/468>`_).


Version 2.0.29
---------------

Released on December 4th 2018.

- Improve user feedback for incorrect specific locus (`#459
  <https://github.com/mutalyzer/mutalyzer/pull/459>`_).
- Improve some position converter warning (`#458
  <https://github.com/mutalyzer/mutalyzer/pull/458>`_).
- Update requirements (`#457
  <https://github.com/mutalyzer/mutalyzer/pull/457>`_).
- Make API more consistent between UDs and NCs (`#456
  <https://github.com/mutalyzer/mutalyzer/pull/456>`_).
- Fix internal server error on IVS variant (`#452
  <https://github.com/mutalyzer/mutalyzer/pull/452>`_).


Version 2.0.28
---------------

Released on June 11th 2018.

- Support for LRG XML schema version 1.9 (`#449
  <https://github.com/mutalyzer/mutalyzer/pull/449>`_).


Version 2.0.27
---------------

Released on May 25th 2018.

- Accept variant descriptions with NC files as references (`#445
  <https://github.com/mutalyzer/mutalyzer/pull/445>`_).
- Removed support for pending LRG reference files (`#444
  <https://github.com/mutalyzer/mutalyzer/pull/444>`_).
- Add link to wiki page with differences between Mutalyzer and HGVS (`#443
  <https://github.com/mutalyzer/mutalyzer/pull/443>`_).
- Fix for batch processor crash due to NCBI server timeout (`#442
  <https://github.com/mutalyzer/mutalyzer/pull/442>`_).
- Update genbank retriever parameters according to NCBI recommendations (`#439
  <https://github.com/mutalyzer/mutalyzer/pull/439>`_).
- Update getGeneLocation` to use standard strategy to select mapping (`#437
  <https://github.com/mutalyzer/mutalyzer/pull/437>`_).
- Add link to website changelog (`#434
  <https://github.com/mutalyzer/mutalyzer/pull/434>`_).


Version 2.0.26
---------------

Released on July 19th 2017.

- Description-extractor dependency updated to version 2.35 (`#429
  <https://github.com/mutalyzer/mutalyzer/pull/429>`_).
- Fix for negative cCDSStop due to wrong transcript to protein link (`#430
  <https://github.com/mutalyzer/mutalyzer/issues/430>`_).


Version 2.0.25
---------------

Released on May 17th 2017.

- Fix for batch processor crash when trying to alter an "item" column of the
  "batch_queue_items" database table (`#426
  <https://github.com/mutalyzer/mutalyzer/pull/426>`_).

Version 2.0.24
---------------

Released on April 12th 2017.

- Fix for SNP converter crash when called with 'rs0' as parameter.
  SNP converter displays now more warning messages (`#419
  <https://github.com/mutalyzer/mutalyzer/issues/419>`_).


Version 2.0.23
---------------

Released on November 9th 2016.

- Fix for internal server error in the position converter triggered by
  non-variant description (`#333
  <https://github.com/mutalyzer/mutalyzer/issues/333>`_).


Version 2.0.22
--------------

Released on September 20th 2016.

- Biopython dependency updated to version 1.68 (`#411
  <https://github.com/mutalyzer/mutalyzer/pull/411>`_).


Version 2.0.21
--------------

Released on June 24th 2016.

- Fix typo in batch SNP converter, causing all batch jobs to stall (`#409
  <https://github.com/mutalyzer/mutalyzer/pull/409>`_).


Version 2.0.20
--------------

Released on June 22th 2016.

- Accept accession number as transcript selector (`#405
  <https://github.com/mutalyzer/mutalyzer/pull/405>`_).
- Add legend to the output of runMutalyzer service (`#404
  <https://github.com/mutalyzer/mutalyzer/pull/404>`_).
- Remove support for GI numbers (`#396
  <https://github.com/mutalyzer/mutalyzer/pull/396>`_).


Version 2.0.19
--------------

Released on June 9th 2016.

- Fix reference file loading by URL (`#383
  <https://github.com/mutalyzer/mutalyzer/pull/383>`_).
- Consider accession version in `getGeneName` webservice method (`#390
  <https://github.com/mutalyzer/mutalyzer/pull/390>`_).


Version 2.0.18
--------------

Released on May 23rd 2016.

- Fixed installation on CentOS 6 (`#358
  <https://github.com/mutalyzer/mutalyzer/pull/358>`_).
- Auto focus primary form input on page load (`#363
  <https://github.com/mutalyzer/mutalyzer/pull/363>`_).
- Check optional argument for ``dup`` (`#369
  <https://github.com/mutalyzer/mutalyzer/pull/369>`_).
- Fixes for chromosome slicing by gene symbol (`#371
  <https://github.com/mutalyzer/mutalyzer/pull/371>`_).


Version 2.0.17
--------------

Released on March 24th 2016.

- Position converter now checks for selection of gene/transcript (`#346
  <https://github.com/mutalyzer/mutalyzer/pull/346>`_).

mutalyzer/config.py
"""
Module for reading the configuration values from a configuration file.

All communication with this module should be done by using the get function
which returns a configuration value, given a name.

Reading the configuration file is implemented lazily and as such done upon the
first call to the get function.

Configuration is read from the file specified by the `MUTALYZER_SETTINGS`
environment variable, or `mutalyzer.conf` in the current directory if it is
not set.
"""


import os

from configobj import ConfigObj

from mutalyzer.util import singleton


class ConfigurationError(Exception):
    """
    Raised when a configuration file cannot be read.
    """
    pass


def get(name):
    """
    Get a configuration value by name.

    :arg name: Name for the configuration value.
    :type name: string

    :raise ConfigurationError: If configuration value could not be read.
    """
    return _Config().get(name)


@singleton
class _Config():
    """
    Read the configuration file and provide access to its values.

    Please note the limitations from the use of the @singleton decorator as
    described in its docstring.
    """
    def __init__(self):
        """
        Initialise the class with variables read from the configuration
        file.

        Configuration values are read from the file specified by the
        `MUTALYZER_SETTINGS` environment variable, or `mutalyzer.conf` in the
        current directory if it is not set.

        :raises ConfigurationError: If configuration could not be read.
        """
        filename = os.environ.get('MUTALYZER_SETTINGS', 'mutalyzer.conf')
        config = self._load_config(filename)

        # We define default values for many configuration settings (except for
        # some that are mandatory for the user to define, i.e. those in the
        # extras/config.user.example file).
        # Todo: Do not duplicate default values here and in the example config
        #     file template.
        config.setdefault('cachesize', 50)
        config.setdefault('maxDldSize', 10)
        config.setdefault('minDldSize', 512)
        config.setdefault('lrgurl', 'ftp://ftp.ebi.ac.uk/pub/databases/lrgex/')
        config.setdefault('internalDb', 'mutalyzer')
        config.setdefault('dbNames', ['hg18', 'hg19', 'mm10'])
        config.setdefault('defaultDb', 'hg19')
        config.setdefault('LocalMySQLuser', 'mutalyzer')
        config.setdefault('LocalMySQLhost', 'localhost')
        config.setdefault('autoReconnect', False)
        config.setdefault('proteinLinkLifetime', 30)
        config.setdefault('proteinLinkNoneLifetime', 5)
        config.setdefault('datestring', '%Y-%m-%d %H:%M:%S')
        config.setdefault('loglevel', 3)
        config.setdefault('outputlevel', 1)
        config.setdefault('debug', True)
        config.setdefault('flanksize', 25)
        config.setdefault('maxvissize', 25)
        config.setdefault('flankclipsize', 6)
        config.setdefault('mailFrom', 'noreply@humgen.nl')
        config.setdefault('mailSubject', 'Result of Mutalyzer batch check.')
        config.setdefault('resultsDir', config['cache'])
        config.setdefault('PIDfile', '/var/run/mutalyzer/mutalyzer-batchd.pid')
        config.setdefault('batchInputMaxSize', 5)
        config.setdefault('nameCheckOutHeader',
            ['Input', 'Errors | Messages', 'AccNo', 'Genesymbol', 'Variant',
             'Reference Sequence Start Descr.', 'Coding DNA Descr.',
             'Protein Descr.', 'GeneSymbol Coding DNA Descr.',
             'GeneSymbol Protein Descr.', 'Genomic Reference',
             'Coding Reference', 'Protein Reference', 'Affected Transcripts',
             'Affected Proteins', 'Restriction Sites Created',
             'Restriction Sites Deleted'])
        config.setdefault('syntaxCheckOutHeader', ['Input', 'Status'])
        config.setdefault('positionConverterOutHeader',
            ['Input Variant', 'Errors', 'Chromosomal Variant', 'Coding Variant(s)'])
        config.setdefault('snpConverterOutHeader',
            ['Input Variant', 'HGVS description(s)', 'Errors | Messages'])
        config.setdefault('bufSize', 32768)
        config.setdefault('header', ['AccNo', 'Genesymbol', 'Mutation'])
        config.setdefault('threshold', 0.05)
        config.setdefault('spliceAlarm', 2)
        config.setdefault('spliceWarn', 5)
        config.setdefault('piwik', False)
        config.setdefault('piwikBase', 'https://piwik.example.com')
        config.setdefault('piwikSite', 1)

        try:
            # We explicitely read all configuration values ad store them in
            # our own dictionary. This makes sure we notice missing or
            # incorrect values upon instantiation.

            # A few 'special' values.
            self._values = {'autoReconnect': config.as_bool('autoReconnect'),
                            'debug':         config.as_bool('debug'),
                            'piwik':         config.as_bool('piwik'),
                            'threshold':     config.as_float('threshold')}

            # Simple string values.
            for name in ('email', 'cache', 'lrgurl', 'internalDb', 'dbNames',
                         'LocalMySQLuser', 'LocalMySQLhost', 'log',
                         'datestring', 'mailFrom', 'mailSubject',
                         'resultsDir', 'nameCheckOutHeader', 'defaultDb',
                         'syntaxCheckOutHeader', 'positionConverterOutHeader',
                         'snpConverterOutHeader', 'PIDfile', 'header',
                         'piwikBase'):
                self._values[name] = config[name]

            # Simple integer values.
            for name in ('minDldSize', 'loglevel', 'outputlevel', 'flanksize',
                         'maxvissize', 'flankclipsize', 'bufSize',
                         'spliceAlarm', 'spliceWarn', 'piwikSite',
                         'proteinLinkLifetime', 'proteinLinkNoneLifetime'):
                self._values[name] = config.as_int(name)

            # File sizes (given in megabytes, stored in bytes).
            for name in ('cachesize', 'maxDldSize', 'batchInputMaxSize'):
                self._values[name] = config.as_int(name) * 1048576

        except KeyError as e:
            raise ConfigurationError('Missing configuration value: %s' % e)

    def get(self, name):
        """
        Get a configuration value by name.

        :arg name: Name for the configuration value.
        :type name: string

        :raises ConfigurationError: If given configuration value name does not
          exist.
        """
        try:
            return self._values[name]
        except KeyError:
            raise ConfigurationError('No such configuration value: %s' % name)

    def _load_config(self, filename):
        """
        Create a `ConfigObj` from the configuration in `filename`.
        """
        try:
            return ConfigObj(filename)
        except IOError:
            raise ConfigurationError('Could not open configuration file: %s'
                                     % filename)
        except SyntaxError:
            raise ConfigurationError('Could not parse configuration file: %s'
                                     % filename)
